src/fermilib/circuits/low_depth_trotter_simulation.py

```diff
@@ -0,0 +1,213 @@
+"""Module for low-depth Trotterized simulation in the plane wave dual basis."""
+import functools
+import itertools
+import numpy
+
+from fermilib.circuits._parallel_bubble_sort import (
+    index_of_position_in_1d_array, parallel_bubble_sort)
+from fermilib.ops import FermionOperator
+from fermilib.transforms import jordan_wigner
+from fermilib.utils import count_qubits
+
+from projectq.ops import (C, S, Sdag, X, Y, Z, H, CNOT, Swap, Rx, Ry, Rz, Ph,
+                          All, Measure)
+
+
+def fermionic_reorder(register, input_order, target_order=None):
+    """Fermionically reorder the given register.
+
+    Maps the input Jordan-Wigner order to the output order.
+
+    Args:
+        register (projectq.QuReg): Quantum register to reorder.
+        input_order (list): The initial Jordan-Wigner canonical order.
+        target_order (list): The desired Jordan-Wigner canonical order.
+    """
+    if not target_order:
+        target_order = list(range(len(input_order)))
+
+    key = (index_of_position_in_1d_array, [0])
+
+    input_swaps = parallel_bubble_sort(list(itertools.product(input_order)),
+                                       key, len(input_order))
+    input_swaps = [swap[0] for swap_layer in input_swaps
+                   for swap in swap_layer]
+
+    output_swaps = parallel_bubble_sort(list(itertools.product(target_order)),
+                                        key, len(input_order))
+    output_swaps = [swap[0] for swap_layer in output_swaps
+                    for swap in swap_layer]
+
+    # Invert the output swaps to go from the input to the target.
+    swaps = input_swaps + output_swaps[::-1]
+
+    # Swap adjacent modes for all swaps in both rounds.
+    for mode in swaps:
+        C(Z) | (register[mode], register[mode + 1])
+        Swap | (register[mode], register[mode + 1])
+
+
+def special_F_adjacent(register, qubit_index, xx_yy_angle, zz_angle):
+    """Apply the 'special F' (fermionic swap, XX+YY evolution, and ZZ
+    evolution) gate to the register on qubit_index and qubit_index + 1.
+
+    Args:
+        register (projectq.QuReg): The register to apply the gate to.
+        qubit_index (integer): The left qubit to act on.
+        xx_yy_angle (float): The angle for evolution under XX+YY.
+        zz_angle (float): The angle for evolution under ZZ.
+    """
+    Rx(numpy.pi / 2.) | register[qubit_index]
+    CNOT | (register[qubit_index], register[qubit_index + 1])
+    Rx(xx_yy_angle) | register[qubit_index]
+    Ry(xx_yy_angle) | register[qubit_index + 1]
+    CNOT | (register[qubit_index + 1], register[qubit_index])
+    Rx(-numpy.pi / 2.) | register[qubit_index + 1]
+    Rz(zz_angle) | register[qubit_index + 1]
+    Sdag | register[qubit_index + 1]
+    CNOT | (register[qubit_index], register[qubit_index + 1])
+    S | register[qubit_index]
+    S | register[qubit_index + 1]
+
+
+def dual_basis_trotter_step(register, hamiltonian, input_ordering=None,
```

Comment:
I think we should name this something like "simulation_gate_trotter_step" because the dual basis trotter step can also be implemented with the FFFT.